src/adventure/adventuretracker.cpp

```diff
@@ -0,0 +1,164 @@
+#include <QJsonDocument>
+#include <QJsonObject>
+
+#include "adventuretracker.h"
+#include "parser/parserutils.h"
+
+#include <QDebug>
+
+AdventureTracker::AdventureTracker(GameObserver &observer, QObject *const parent)
+    : QObject{parent}
+    , m_observer{observer}
+{
+    connect(&m_observer,
+            &GameObserver::sig_sentToUserText,
+            this,
+            &AdventureTracker::slot_onUserText);
+
+    connect(&m_observer,
+            &GameObserver::sig_sentToUserGmcp,
+            this,
+            &AdventureTracker::slot_onUserGmcp);
+}
+
+AdventureTracker::~AdventureTracker() {}
+
+void AdventureTracker::slot_onUserText(const QByteArray &ba)
+{
+    QString str = QString::fromLatin1(ba).trimmed();
+    ParserUtils::removeAnsiMarksInPlace(str);
+
+    // We keep track of the N most recent lines so that we can parse with the
+    // context of a sliding window. Array indices are interpeted backwards,
+    // so [0] is the last line received, [1] is the second-to-last, etc.
+    // Must create our own QString copy for this, since the `ba` passed above
+    // is reused. So shfit the array and add a copy.
+    for (size_t i = m_lastLines.size(); i > 0; i--) {
+        m_lastLines[i] = m_lastLines[i - 1];
+    }
+    m_lastLines[0] = new QString(str);
+
+    parseIfKillAndXP();
+
+    parseIfAchievedSomething();
+
+    if (str.contains("You gain a level!")) {
```

Comment:
This might trigger on a whois. Why not do an exact match?

Reply:
Good catch! I have since updated to match only on the start of a line (so indexOf == 0). But in the mid-term, I should be inferring a level change from GMCP and not need to do this check.

Ultimately, I hope to do as much of this as possible from GMCP (thus significantly mitigating the "whois" issue), and probably the only thing I will parse from text will be the Something is Dead R.I.P. and XP attribution